### ISSUE TITLE
R2.pred for phylolm

### ISSUE THREAD

**Issue by arives** [arives/rr2#3]
I tried a version of R2.pred for phylolm objects in which predictions were taken as the estimated value of the trait at the closest node to the tip. This is consistent with R2.pred for lmerMod objects. This can be done using the ace {ape} or directly using the conditional multivariate normal. The code for ace {ape} is

phy.f$edge <- phy.f$edge[order(phy.f$edge[,2]),]
R <- Y - stats::fitted(mod)
Rnode.f <- ape::ace(x=R, phy=phy.f, method="ML")$ace
nodes.f <- phy.f$edge[is.element(phy.f$edge[,2],1:n),1]
Rhat.f <- Rnode.f[match(nodes.f,names(Rnode.f))]
Yhat.f <- as.numeric(stats::fitted(mod) + Rhat.f)
SSE.pred <- var(Y - Yhat.f)

I wrote code for this, but the values are generally much higher than those for R2.resid and R2.lik, so I don't think this is a good approach in practice, even though it makes sense in theory (to match R2.pred.lmerMod)

**Comment by daijiang** [comment]
@arives is this solved?? Do we still need `predict(mod.lmer, re.form = NA)`?

**Comment by daijiang** [comment]
In current way, [this](https://github.com/arives/rr2/blob/944a1dc4d1383c9124610ccda6f79385ebeddfd6/R/r2_pred.R#L103-L108) `if else` is not necessary. If this is the case, then `R2.pred.lmerMod` reduce to `R2.pred.lm`.

**Comment by arives** [comment]
Daijiang,

I'm still working on this. I'll let you know when I'm done. Most of the time I work on things outside of github, but there are some things I need to do to figure out how to call functions.

Cheers, Tony

**Comment by daijiang** [comment]
Email me if you need help.

**Comment by arives** [comment]
I'm going down the rabbit hole of rethinking how to do R2.pred and R2.resid for continuous models. Hellllllp…

Seriously, I'll let you know if anything comes up. I want to finish today.


______________
Anthony R. Ives
UW-Madison
459 Birge Hall
608-262-1519

**Comment by arives** [comment]
I looked at this a little more, to reassess whether R2.pred.phylolm used the best way of making predictions. Right now I think I'm going to change R2.pred.phylolm to use the same approach used for R2.communityPGLMM.gaussian; it is more elegant from a mathematical point of view, and it matches R2.pred.lmerMod. This might make the "tip_rm" option for R2.communityPLGMM.gaussian unnecessary, although it is nice to have.

One thing about R2.pred.lmerMod with "nearest_node" (and the similar version for communityPGLMM and phylolm) is that it doesn't behave all that well for small n. It is more variable than the "tip_rm" option, which is interesting, since I would have thought it was the other way around. But I did double check with this lmer's fitted() values. The problems are really due to the R2 associated with fixed effects; when fixed effects are simulated as zero, then the partial R2 for them aren't very good.

Changing R2.pred.phylolm to the "tip_rm" version will give higher values than R2.resid and R2.lik. I'm not happy about that. Nothing is perfect.

This is one of those annoying problems where the decision isn't so much mathematical but subjective. The issue is matching R2.pred.lmerMod mathematically or R2.resid.phylolm quantitatively. Given that R2.resid.phylolm is a little ad hoc (standardizing phylogenies to have the same total edge lengths), this has me leaning towards matching R2.pred.lmerMod.
